File: rust/tests/transactions_decorator_tests.rs
//! Integration tests for the optional transactions decorator (Rust side).
//!
//! Mirrors the C# `TransactionsDecoratorTests` and exercises the
//! requirements R1–R10 of issue #94.

use anyhow::Result;
use link_cli::transactions::DoubletLink;
use link_cli::{
    CommitMode, LogRetentionPolicy, NamedTypesDecorator, TransactionsDecorator, Transition,
    TransitionKind,
};
use std::path::PathBuf;
use tempfile::NamedTempFile;

fn make_tx() -> (TransactionsDecorator, Vec<NamedTempFile>) {
    let data_file = NamedTempFile::new().expect("create temp file");
    let log_file = NamedTempFile::new().expect("create temp file");
    let data_links =
        NamedTypesDecorator::new(data_file.path(), false).expect("open data links");
    let log_links = NamedTypesDecorator::new(log_file.path(), false).expect("open log links");
    let tx = TransactionsDecorator::new(
        data_links,
        log_links,
        LogRetentionPolicy::default(),
        CommitMode::default(),
        false,
    )
    .expect("open transactions decorator");
    // Keep the temp files alive for the duration of the test.
    (tx, vec![data_file, log_file])
}

#[test]
fn auto_transaction_records_create_and_update() -> Result<()> {
    let (mut tx, _guards) = make_tx();

    let created = tx.create_and_update(0, 0)?;
    assert_ne!(0, created);

    let log = tx.log();
    assert_eq!(2, log.len(), "create_and_update must record two transitions");
    assert_eq!(TransitionKind::Create, log[0].kind);
    assert_eq!(TransitionKind::Update, log[1].kind);
    assert_eq!(created, log[0].after.index);
    Ok(())
}

#[test]
fn rollback_undoes_create() -> Result<()> {
    let (mut tx, _guards) = make_tx();

    tx.begin_transaction()?;
    let created = tx.create_and_update(0, 0)?;
    assert!(tx.exists(created));
    tx.rollback()?;

    assert!(!tx.exists(created), "rolled-back create must remove the link");
    Ok(())
}

#[test]
fn commit_persists_create() -> Result<()> {
    let (mut tx, _guards) = make_tx();

    tx.begin_transaction()?;
    let created = tx.create_and_update(0, 0)?;
    tx.commit()?;

    assert!(tx.exists(created));
    assert_eq!(tx.last_logged_sequence(), tx.applied_sequence());
    Ok(())
}

#[test]
fn rollback_undoes_update() -> Result<()> {
    let (mut tx, _guards) = make_tx();

    let a = tx.create_and_update(0, 0)?;
    let b = tx.create_and_update(0, 0)?;
    let c = tx.create_and_update(0, 0)?;

    tx.begin_transaction()?;
    tx.update(c, a, b)?;
    let updated = tx.get(c).copied().unwrap();
    assert_eq!(a, updated.source);
    assert_eq!(b, updated.target);
    tx.rollback()?;

    let after_rollback = tx.get(c).copied().unwrap();
    assert_eq!(c, after_rollback.index);
    assert_eq!(0, after_rollback.source);
    assert_eq!(0, after_rollback.target);
    Ok(())
}

#[test]
fn rollback_undoes_delete() -> Result<()> {
    let (mut tx, _guards) = make_tx();

    let a = tx.create_and_update(0, 0)?;
    let b = tx.create_and_update(0, 0)?;
    let c = tx.create_and_update(0, 0)?;
    tx.update(c, a, b)?;

    tx.begin_transaction()?;
    tx.delete(c)?;
    assert!(!tx.exists(c));
    tx.rollback()?;

    assert!(tx.exists(c), "delete must be restored by rollback");
    let restored = tx.get(c).copied().unwrap();
    assert_eq!(a, restored.source);
    assert_eq!(b, restored.target);
    Ok(())
}

#[test]
fn sized_retention_drops_oldest_after_applied() -> Result<()> {
    let (mut tx, _guards) = make_tx();
    tx.set_retention_policy(LogRetentionPolicy::Sized {
        max_transitions: 3,
    });

    for _ in 0..5 {
        tx.create_and_update(0, 0)?;
    }

    assert!(
        tx.log().len() as u64 <= 3,
        "sized retention must cap log length; got {}",
        tx.log().len()
    );
    Ok(())
}

#[test]
fn chunked_retention_archives_oldest() -> Result<()> {
    let archive_dir = std::env::temp_dir().join(format!(
        "tx-archive-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos()
    ));
    let _ = std::fs::remove_dir_all(&archive_dir);

    {
        let (mut tx, _guards) = make_tx();
        tx.set_retention_policy(LogRetentionPolicy::Chunked {
            chunk_size: 2,
            archive_directory: archive_dir.clone(),
        });

        for _ in 0..4 {
            tx.create_and_update(0, 0)?;
        }

        assert!(archive_dir.exists(), "archive directory must be created");
        let files: Vec<_> = std::fs::read_dir(&archive_dir)?
            .filter_map(|e| e.ok())
            .filter(|e| {
                e.file_name()
                    .to_string_lossy()
                    .starts_with("transitions-chunk-")
            })
            .collect();
        assert!(!files.is_empty(), "chunked retention must archive files");
    }
    let _ = std::fs::remove_dir_all(&archive_dir);
    Ok(())
}

#[test]
fn retention_policy_parses_specs() {
    assert!(matches!(
        LogRetentionPolicy::parse("infinite").unwrap(),
        LogRetentionPolicy::Infinite
    ));
    assert!(matches!(
        LogRetentionPolicy::parse("sized:1000").unwrap(),
        LogRetentionPolicy::Sized {
            max_transitions: 1000
        }
    ));
    match LogRetentionPolicy::parse("chunked:500:/tmp/x").unwrap() {
        LogRetentionPolicy::Chunked {
            chunk_size,
            archive_directory,
        } => {
            assert_eq!(chunk_size, 500);
            assert_eq!(archive_directory, PathBuf::from("/tmp/x"));
        }
        _ => panic!("expected Chunked"),
    }
    assert!(LogRetentionPolicy::parse("garbage").is_err());
}

#[test]
fn transition_round_trips_through_serialize() {
    let t = Transition {
        transaction_id: 0xabcdef1234567890u128,
        sequence: 42,
        timestamp_ms: 1234567890,
        kind: TransitionKind::Update,
        before: DoubletLink::new(1, 2, 3),
        after: DoubletLink::new(1, 4, 5),
    };

    let parsed = Transition::try_parse(&t.serialize()).unwrap();
    assert_eq!(t, parsed);
}

#[test]
fn async_commit_marks_applied() -> Result<()> {
    // Rust port runs sync; the contract that `applied_sequence` reaches
    // `last_logged_sequence` after commit still holds.
    let (mut tx, _guards) = make_tx();
    tx.set_commit_mode(CommitMode::Async);

    tx.begin_transaction()?;
    let created = tx.create_and_update(0, 0)?;
    tx.commit()?;

    assert_eq!(tx.last_logged_sequence(), tx.applied_sequence());
    assert!(tx.exists(created));
    Ok(())
}

#[test]
fn no_behaviour_change_when_not_opted_in() -> Result<()> {
    // R8: bare NamedTypesDecorator behaves identically whether or not
    // the TransactionsDecorator is wrapped above it.
    let data_file = NamedTempFile::new()?;
    let mut data_links = NamedTypesDecorator::new(data_file.path(), false)?;
    let id = data_links.get_or_create(0, 0);
    assert!(data_links.exists(id));
    Ok(())
}

#[test]
fn recovery_reapplies_committed_transitions() -> Result<()> {
    // After dropping the decorator and reopening with the same store
    // files, the recovery protocol should rebuild the log and restore
    // committed state (R6).
    let data_file = NamedTempFile::new()?;
    let log_file = NamedTempFile::new()?;
    let data_path = data_file.path().to_path_buf();
    let log_path = log_file.path().to_path_buf();

    let id = {
        let data_links = NamedTypesDecorator::new(&data_path, false)?;
        let log_links = NamedTypesDecorator::new(&log_path, false)?;
        let mut tx = TransactionsDecorator::new(
            data_links,
            log_links,
            LogRetentionPolicy::default(),
            CommitMode::default(),
            false,
        )?;
        tx.begin_transaction()?;
        let id = tx.create_and_update(0, 0)?;
        tx.commit()?;
        tx.save()?;
        id
    };

    let data_links = NamedTypesDecorator::new(&data_path, false)?;
    let log_links = NamedTypesDecorator::new(&log_path, false)?;
    let tx = TransactionsDecorator::new(
        data_links,
        log_links,
        LogRetentionPolicy::default(),
        CommitMode::default(),
        false,
    )?;

    assert!(tx.exists(id), "committed link must survive reopen");
    assert!(tx.last_logged_sequence() >= 2);
    assert!(tx.applied_sequence() >= tx.last_logged_sequence());
    Ok(())
}

#[test]
fn make_transitions_database_filename_returns_sibling_path() {
    let path = TransactionsDecorator::make_transitions_database_filename("/var/data/db.links");
    assert_eq!(path, PathBuf::from("/var/data/db.transitions.links"));

    let path = TransactionsDecorator::make_transitions_database_filename("db.links");
    assert_eq!(path, PathBuf::from("db.transitions.links"));
}

#[test]
fn nested_transactions_are_rejected() -> Result<()> {
    let (mut tx, _guards) = make_tx();
    tx.begin_transaction()?;
    let result = tx.begin_transaction();
    assert!(result.is_err(), "nested transactions must be rejected");
    tx.rollback()?;
    Ok(())
}

File: csharp/Foundation.Data.Doublets.Cli.Tests/TransactionsDecoratorTests.cs
using Platform.Data;
using Platform.Data.Doublets;

using DoubletLink = Platform.Data.Doublets.Link<uint>;

namespace Foundation.Data.Doublets.Cli.Tests.Tests
{
    public class TransactionsDecoratorTests
    {
        [Fact]
        public void AutoTransactionRecordsCreateAndUpdate()
        {
            // CreateAndUpdate is an extension that calls Create then Update on
            // the doublets store. Each emits a transition.
            RunWithTransactions((tx, _) =>
            {
                var created = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                Assert.NotEqual(tx.Constants.Null, created);

                var log = tx.Log;
                Assert.Equal(2, log.Count);
                Assert.Equal(TransitionKind.Create, log[0].Kind);
                Assert.Equal(TransitionKind.Update, log[1].Kind);
                Assert.Equal(created, log[0].After.Index);
            });
        }

        [Fact]
        public void RollbackUndoesCreate()
        {
            RunWithTransactions((tx, _) =>
            {
                uint created;
                using (var transaction = tx.BeginTransaction())
                {
                    created = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                    Assert.True(tx.Exists(created));
                    transaction.Rollback();
                }

                Assert.False(tx.Exists(created), "Rolled-back create must remove the link.");
            });
        }

        [Fact]
        public void DisposeWithoutCommitRollsBack()
        {
            RunWithTransactions((tx, _) =>
            {
                uint created;
                using (var transaction = tx.BeginTransaction())
                {
                    created = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                }

                Assert.False(tx.Exists(created), "Disposing an open transaction must rollback (R10).");
            });
        }

        [Fact]
        public void CommitPersistsCreate()
        {
            RunWithTransactions((tx, _) =>
            {
                uint created;
                using (var transaction = tx.BeginTransaction())
                {
                    created = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                    transaction.Commit();
                }

                Assert.True(tx.Exists(created));
                Assert.Equal(tx.LastLoggedSequence, tx.AppliedSequence);
            });
        }

        [Fact]
        public void RollbackUndoesUpdate()
        {
            RunWithTransactions((tx, _) =>
            {
                var a = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                var b = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                var c = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);

                using (var transaction = tx.BeginTransaction())
                {
                    tx.Update(
                new DoubletLink(c, tx.Constants.Any, tx.Constants.Any),
                new DoubletLink(c, a, b),
                null);
                    var updated = new DoubletLink(tx.GetLink(c));
                    Assert.Equal(a, updated.Source);
                    Assert.Equal(b, updated.Target);
                    transaction.Rollback();
                }

                var afterRollback = new DoubletLink(tx.GetLink(c));
                Assert.Equal(c, afterRollback.Index);
                Assert.Equal(tx.Constants.Null, afterRollback.Source);
                Assert.Equal(tx.Constants.Null, afterRollback.Target);
            });
        }

        [Fact]
        public void RollbackUndoesDelete()
        {
            RunWithTransactions((tx, _) =>
            {
                var a = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                var b = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                var c = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                tx.Update(
            new DoubletLink(c, tx.Constants.Any, tx.Constants.Any),
            new DoubletLink(c, a, b),
            null);

                using (var transaction = tx.BeginTransaction())
                {
                    tx.Delete(new DoubletLink(c, tx.Constants.Any, tx.Constants.Any), null);
                    Assert.False(tx.Exists(c));
                    transaction.Rollback();
                }

                Assert.True(tx.Exists(c), "Delete must be restored by rollback.");
                var restored = new DoubletLink(tx.GetLink(c));
                Assert.Equal(a, restored.Source);
                Assert.Equal(b, restored.Target);
            });
        }

        [Fact]
        public void SizedRetentionDropsOldestAfterApplied()
        {
            RunWithTransactions((tx, _) =>
            {
                tx.RetentionPolicy = new LogRetentionPolicy.Sized(MaxTransitions: 3);
                for (var i = 0; i < 5; i++)
                {
                    tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                }

                Assert.True(tx.Log.Count <= 3, $"Sized retention must cap log length; got {tx.Log.Count}.");
            });
        }

        [Fact]
        public void ChunkedRetentionArchivesOldest()
        {
            var archiveDir = Path.Combine(Path.GetTempPath(), $"tx-archive-{Guid.NewGuid():N}");
            try
            {
                RunWithTransactions((tx, _) =>
                {
                    tx.RetentionPolicy = new LogRetentionPolicy.Chunked(ChunkSize: 2, ArchiveDirectory: archiveDir);
                    for (var i = 0; i < 4; i++)
                    {
                        tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                    }

                    Assert.True(Directory.Exists(archiveDir));
                    var files = Directory.EnumerateFiles(archiveDir, "transitions-chunk-*.log").ToList();
                    Assert.NotEmpty(files);
                });
            }
            finally
            {
                if (Directory.Exists(archiveDir)) Directory.Delete(archiveDir, recursive: true);
            }
        }

        [Fact]
        public void RetentionPolicyParsesSpecs()
        {
            Assert.IsType<LogRetentionPolicy.Infinite>(LogRetentionPolicy.Parse("infinite"));
            Assert.IsType<LogRetentionPolicy.Sized>(LogRetentionPolicy.Parse("sized:1000"));
            Assert.IsType<LogRetentionPolicy.Chunked>(LogRetentionPolicy.Parse("chunked:500:/tmp/x"));
            Assert.Throws<ArgumentException>(() => LogRetentionPolicy.Parse("garbage"));
        }

        [Fact]
        public void TransitionRoundTripsThroughSerialize()
        {
            var t = new Transition(
              Guid.NewGuid(),
              Sequence: 42,
              Timestamp: DateTimeOffset.FromUnixTimeMilliseconds(1234567890000),
              Kind: TransitionKind.Update,
              Before: new DoubletLink(1, 2, 3),
              After: new DoubletLink(1, 4, 5));

            Assert.True(Transition.TryParse(t.Serialize(), out var parsed));
            Assert.Equal(t, parsed);
        }

        [Fact]
        public void AsyncCommitEventuallyMarksApplied()
        {
            RunWithTransactions((tx, _) =>
            {
                tx.CommitMode = CommitMode.Async;
                uint created;
                using (var transaction = tx.BeginTransaction())
                {
                    created = tx.CreateAndUpdate(tx.Constants.Null, tx.Constants.Null);
                    transaction.CommitAsync().GetAwaiter().GetResult();
                }

                // Allow background worker time to drain.
                var deadline = DateTime.UtcNow.AddSeconds(5);
                while (tx.AppliedSequence < tx.LastLoggedSequence && DateTime.UtcNow < deadline)
                {
                    Thread.Sleep(50);
                }
                Assert.Equal(tx.LastLoggedSequence, tx.AppliedSequence);
                Assert.True(tx.Exists(created));
            });
        }

        [Fact]
        public void NoBehaviourChangeWhenNotOptedIn()
        {
            // Acceptance for R8: bare NamedTypesDecorator behaves identically
            // whether or not TransactionsDecorator is wrapped above it.
            var dataFile = Path.GetTempFileName();
            NamedTypesDecorator<uint>? dataLinks = null;
            try
            {
                dataLinks = new NamedTypesDecorator<uint>(dataFile);
                var created = dataLinks.CreateAndUpdate(dataLinks.Constants.Null, dataLinks.Constants.Null);
                Assert.True(dataLinks.Exists(created));
            }
            finally
            {
                dataLinks?.Dispose();
                Cleanup(dataFile);
                Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(dataFile));
            }
        }

        private static void RunWithTransactions(Action<TransactionsDecorator, NamedTypesDecorator<uint>> action)
        {
            var dataFile = Path.GetTempFileName();
            var logFile = Path.GetTempFileName();
            NamedTypesDecorator<uint>? dataLinks = null;
            NamedTypesDecorator<uint>? logLinks = null;
            TransactionsDecorator? tx = null;
            try
            {
                dataLinks = new NamedTypesDecorator<uint>(dataFile);
                logLinks = new NamedTypesDecorator<uint>(logFile);
                tx = new TransactionsDecorator(dataLinks, logLinks);
                action(tx, dataLinks);
            }
            finally
            {
                tx?.Dispose();
                dataLinks?.Dispose();
                logLinks?.Dispose();
                Cleanup(dataFile);
                Cleanup(logFile);
                Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(dataFile));
                Cleanup(NamedTypesDecorator<uint>.MakeNamesDatabaseFilename(logFile));
            }
        }

        private static void Cleanup(string path)
        {
            if (File.Exists(path)) File.Delete(path);
        }
    }
}
